src/libsyntax/parse/parser.rs

```diff
@@ -5131,6 +5183,34 @@ impl<'a> Parser<'a> {
                     span: lo.to(self.prev_span),
                 });
                 seen_binding = true;
+            } else if self.check_const_arg() {
+                // TODO(const_generics): Numbers and expressions in braces are
+                // easy to check for, but bare Idents for `const`s or const
+                // params are indistinguishable from Idents for type params.
+                // There are two possibilities I can see to resolving the
+                // parsing problem. Either mark all const args with const (or
+                // something else, it doesn't really matter) or merge const args
+                // and type args together at the AST level and separate them
+                // later on, probably when lowering to the HIR.
+                //
+                // For now we just ignore Idents pointing to consts so we can
+                // get everything else working.
+                let expr = if let token::OpenDelim(token::Brace) = self.token {
+                    let body_lo = self.span;
+                    self.parse_block_expr(None, body_lo, BlockCheckMode::Default, ThinVec::new())?
+                } else if self.token.can_begin_literal_or_bool() {
+                    let lit = self.parse_lit()?;
+                    self.mk_expr(lit.span, ExprKind::Lit(P(lit)), ThinVec::new())
+                } else {
+                    unreachable!()
```

Comment:
This will hit unreachable on `S<-1>`.
I'd just use `parse_expr` if `self.check_const_arg()` returns true.

(Unless you still have plans regarding turbofish, in this case a comment here explaining "why not `parse_expr`" would be appropriate.)

Reply:
Ah, this was an `unreachable` I missed when filling things out. `parse_expr` should do fine for now.